recirq/quantum_chess/quantum_board.py

```diff
@@ -311,39 +323,123 @@ def sample(self, num_samples: int) -> List[int]:
 
+    def _caching_supported(self, m: move.Move):
+        """Checks if caching is supported for this move."""
+
+        # Caching is supported for a split jump from one full square to two empty squares.
+        if m.move_type == enums.MoveType.SPLIT_JUMP and nth_bit_of(square_to_bit(m.source), self.full_squares)\
+            and nth_bit_of(square_to_bit(m.target), self.empty_squares)\
+            and nth_bit_of(square_to_bit(m.target2), self.empty_squares):
+            return True
+        return False
+
+    def cache_results(self, cache_key: CacheKey):
+        if cache_key in self.cache:
+            return
+        if cache_key.move_type == enums.MoveType.SPLIT_JUMP:
+            helper_board = CirqBoard(DEFAULT_CHESS_INIT_STATE, self.sampler,
+                                     self.device, self.error_mitigation,
+                                     self.noise_mitigation,
+                                     self.transformer if self.device else None)
+            sample_jump_move = move.Move('b1',
+                                         'c3',
+                                         target2='a3',
+                                         move_type=enums.MoveType.SPLIT_JUMP,
+                                         move_variant=enums.MoveVariant.BASIC)
+            helper_board.do_move(sample_jump_move)
+            probs = helper_board.get_probability_distribution(
+                cache_key.repetitions, use_cache=False)
+            self.cache[cache_key] = {
+                "source": 0,
+                "target": probs[square_to_bit(sample_jump_move.target)],
+                "target2": probs[square_to_bit(sample_jump_move.target2)]
+            }
+            self.debug_log += helper_board.debug_log
+
+    @staticmethod
+    def _apply_cache(probability, m, cache_value):
+        new_probability = probability.copy()
+        for k, v in cache_value.items():
+            square = getattr(m, k)
+            new_probability[square_to_bit(square)] = v
+        return new_probability
+
     def get_probability_distribution(self,
-                                     repetitions: int = 1000) -> List[float]:
+                                     repetitions: int = 1000,
+                                     use_cache=True) -> List[float]:
         """Returns the probability of a piece being in each square.
 
         The values are returned as a list in the same ordering as a
         bitboard.
         """
         if self.accumulations_repetitions != repetitions:
-            self._generate_accumulations(repetitions)
+            self._generate_accumulations(repetitions, use_cache)
 
+        # Cache the repetiion and probability corresponding to this move history.
+        if len(self.move_history_probabilities_cache) < len(self.move_history):
+            self.move_history_probabilities_cache.extend([(-1, None)] * (len(self.move_history) - len(self.move_history_probabilities_cache)))
```

Comment:
How about making the length of `self.move_history_probabilities_cache` always match the length of `self.move_history`? Just adding the dummy `(-1, None)` entry whenever the move history changes. Then there will won't need to be logic checking its length.

Reply:
Thanks, done.